File: models/Gpt-3.5-turbo/cpp/code/problems/introductory_problems/Chessboard_and_Queens.cpp

#include <iostream>
using namespace std;

const int N = 8; // size of the chessboard
int count = 0; // variable to store the count of valid placements

bool isSafe(int row, int col, char board[N][N]) {
    // check if there is a queen in the same row or same column
    for (int i = 0; i < N; i++) {
        if (board[row][i] == 'Q' || board[i][col] == 'Q') {
            return false;
        }
    }
    
    // check if there is a queen in the diagonal
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 'Q') {
            return false;
        }
    }
    
    for (int i = row, j = col; i >= 0 && j < N; i--, j++) {
        if (board[i][j] == 'Q') {
            return false;
        }
    }
    
    return true;
}

void solve(int row, char board[N][N]) {
    // if all queens are placed, increment the count
    if (row == N) {
        count++;
        return;
    }
    
    // try placing the queen in each column of the current row
    for (int i = 0; i < N; i++) {
        if (board[row][i] == '.' && isSafe(row, i, board)) {
            board[row][i] = 'Q'; // place the queen
            
            // recursively move to the next row
            solve(row + 1, board);
            
            board[row][i] = '.'; // backtrack and remove the queen
        }
    }
}

int main() {
    char board[N][N];
    
    // read the input chessboard
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> board[i][j];
        }
    }
    
    solve(0, board);
    
    cout << count << endl;
    
    return 0;
}

File: models/Gpt-3.5-turbo/py/code/introductory_problems/Chessboard_and_Queens.py

def is_safe(board, row, col):
    # Check if a queen can be placed at board[row][col]
    
    # Check the same column
    for i in range(row):
        if board[i][col] == 'Q':
            return False
    
    # Check the upper diagonal
    i, j = row, col
    while i >= 0 and j >= 0:
        if board[i][j] == 'Q':
            return False
        i -= 1
        j -= 1
    
    # Check the lower diagonal
    i, j = row, col
    while i >= 0 and j < 8:
        if board[i][j] == 'Q':
            return False
        i -= 1
        j += 1
    
    return True


def count_arrangements(board, row):
    # Base case: If all queens are placed, increment the count
    if row == 8:
        return 1
    
    count = 0
    
    # Try to place a queen in each column of the current row
    for col in range(8):
        if is_safe(board, row, col) and board[row][col] == '.':
            # Place the queen
            board[row][col] = 'Q'
            
            # Recur for the next row
            count += count_arrangements(board, row + 1)
            
            # Remove the queen (backtrack)
            board[row][col] = '.'
    
    return count


# Read the input chessboard
board = [list(input()) for _ in range(8)]

# Count the number of arrangements
arrangements = count_arrangements(board, 0)

# Print the result
print(arrangements)
